resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -617,30 +620,30 @@ void tryToFinishResolve(final DnsServerAddressStream nameServerAddrStream,
                 if (queryLifecycleObserver == NoopDnsQueryLifecycleObserver.INSTANCE) {
                     // If the queryLifecycleObserver has already been terminated we should create a new one for this
                     // fresh query.
-                    query(nameServerAddrStream, nameServerAddrStreamIndex + 1, question, promise);
+                    query(nameServerAddrStream, nameServerAddrStreamIndex + 1, question, promise, cause);
                 } else {
                     query(nameServerAddrStream, nameServerAddrStreamIndex + 1, question, queryLifecycleObserver,
-                          promise);
+                          promise, cause);
                 }
                 return;
             }
 
             queryLifecycleObserver.queryFailed(NAME_SERVERS_EXHAUSTED_EXCEPTION);
 
             // .. and we could not find any A/AAAA records.
-            if (!triedCNAME) {
+            if (cause == null && !triedCNAME) {
```

Comment:
is it possible that the current query failed, but the CNAME query may succeed? If so we may be prematurely giving up on the query.

Reply:
I think if the A / AAA record query failed due of an timeout or IO error we want to not even try the CNAME and just retry or whatever the user wants to do. Not strong on this tho.

Reply:
Isn't it possible the CNAME query may actually try a different DNS server (via the `DnsServerAddressStream`), or be cached?

Reply:
Yes its possible and I think this is even a problem... We should only try CNAME if an A or AAAA record not exists (and this was the case for the failure), not if we failed because of a timeout.

Reply:
Oh I see ... for timeout/cancel this makes sense ... what about transport error though?

Reply:
I think its the same here... any reason why you think this is not the case ?

Reply:
if its a transport failure ... this may not apply to the CNAME query because it may use a different transport, right?